core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -677,9 +684,54 @@ public PlanWithProperties visitFilter(FilterNode node, PreferredProperties prefe
         @Override
         public PlanWithProperties visitTableScan(TableScanNode node, PreferredProperties preferredProperties)
         {
+            // attempt tp push preferred partitioning into the table scan
+            var partitioningProperties = preferredProperties.getGlobalProperties().flatMap(PreferredProperties.Global::getPartitioningProperties);
+            if (isUseTableScanNodePartitioning(session) && partitioningProperties.isPresent()) {
+                Optional<TableScanNode> partitionedNode = pushPartitioningIntoTableScan(plannerContext.getMetadata(), session, node, partitioningProperties.get());
+                if (partitionedNode.isPresent()) {
+                    node = partitionedNode.get();
+                }
+            }
             return new PlanWithProperties(node, deriveProperties(node, ImmutableList.of()));
         }
 
+        private Optional<TableScanNode> pushPartitioningIntoTableScan(Metadata metadata, Session session, TableScanNode node, PreferredProperties.PartitioningProperties partitioningProperties)
+        {
+            verify(node.getAssignments().keySet().containsAll(partitioningProperties.getPartitioningColumns()), "TableScanNode does not read all partitioning columns");
+
+            List<ColumnHandle> partitionColumns;
+            if (partitioningProperties.getPartitioning().isPresent()) {
+                Partitioning partitioning = partitioningProperties.getPartitioning().get();
+                // constant partitioning values cannot be pushed into the table scan
```

Comment:
When we have a combination of constants and partitioning columns, are we not able to take advantage of partitioning ?

Reply:
I'm not sure that its relevant here but I have seen a plan like below
```
Fragment 1 [mmdsbx:HivePartitioningHandle{buckets=32, hiveTypes=[string, int]}]
    CPU: 1.71m, Scheduled: 4.40m, Blocked 8.73m (Input: 44.35s, Output: 51.48s), Input: 19435177 rows (2.30GB); per task: avg.: 19435177.00 std.dev.: 0.00, Output: 14982018 rows (5.32GB)
    Output layout: [root_guid, guid, lvl1_root_guid, concat, concat_4, redundant, cycle, excluded]
    Output partitioning: SINGLE []
    └─ Aggregate[type = FINAL, keys = [root_guid, guid, lvl1_root_guid, concat, concat_4, redundant, cycle, excluded], hash = [$hashvalue]]
       │   Layout: [root_guid:varchar, guid:varchar, lvl1_root_guid:varchar, concat:array(varchar), concat_4:varchar, redundant:boolean, cycle:boolean, excluded:boolean, $hashvalue:bigint]
       └─ LocalExchange[partitioning = HASH, hashColumn = [$hashvalue], arguments = ["root_guid", "guid", "lvl1_root_guid", "concat", "concat_4", "redundant", "cycle", "excluded"]]
          │   Layout: [root_guid:varchar, guid:varchar, lvl1_root_guid:varchar, concat:array(varchar), concat_4:varchar, redundant:boolean, cycle:boolean, excluded:boolean, $hashvalue:bigint]
          └─ Aggregate[type = PARTIAL, keys = [root_guid, guid, lvl1_root_guid, concat, concat_4, redundant, cycle, excluded], hash = [$hashvalue_9]]
             │   Layout: [root_guid:varchar, guid:varchar, lvl1_root_guid:varchar, concat:array(varchar), concat_4:varchar, redundant:boolean, cycle:boolean, excluded:boolean, $hashvalue_9:bigint]
             └─ InnerJoin[criteria = ("guid" = "root_guid_0"), hash = [$hashvalue_5, $hashvalue_6], distribution = PARTITIONED]
                      │   Layout: [root_guid:varchar, guid:varchar, cycle:boolean, excluded:boolean, redundant:boolean, xpath:array(varchar), lvl1_root_guid:varchar]
                      │   Distribution: PARTITIONED
                      │   maySkipOutputDuplicates = true
                      │   dynamicFilterAssignments = {root_guid_0 -> #df_631}
                      ├─ ScanFilterProject[table = mmdsbx:stg_gbt_top_down_expanded_2 buckets=32, filterPredicate = ("level" = 3), dynamicFilters = {"guid" = #df_631}]
                      │      Layout: [root_guid:varchar, guid:varchar, cycle:boolean, excluded:boolean, redundant:boolean, xpath:array(varchar), $hashvalue_5:bigint]
                      │      $hashvalue_5 := combine_hash(bigint '0', COALESCE("$operator$hash_code"("guid"), 0))
                      │      excluded := excluded:boolean:REGULAR
                      │      xpath := xpath:array<string>:REGULAR
                      │      level := level:int:REGULAR
                      │      guid := guid:string:REGULAR
                      │      root_guid := root_guid:string:REGULAR
                      │      cycle := cycle:boolean:REGULAR
                      │      redundant := redundant:boolean:REGULAR
                      │      Input: 11886091 rows (1.72GB), Filtered: 44.99%
                      │      Dynamic filters: 
                      │          - df_631, ALL, collection time=564.27ms
                      └─ LocalExchange[partitioning = HASH, hashColumn = [$hashvalue_6], arguments = ["root_guid_0"]]
                         │   Layout: [root_guid_0:varchar, lvl1_root_guid:varchar, $hashvalue_6:bigint]
                         └─ RemoteSource[sourceFragmentIds = [2]]
                                Layout: [root_guid_0:varchar, lvl1_root_guid:varchar, $hashvalue_7:bigint]

Fragment 2 [SOURCE]
    CPU: 3.27s, Scheduled: 7.72s, Blocked 0.00ns (Input: 0.00ns, Output: 0.00ns), Input: 7549086 rows (532.75MB); per task: avg.: 7549086.00 std.dev.: 0.00, Output: 7549086 rows (597.55MB)
    Output layout: [root_guid_0, lvl1_root_guid, $hashvalue_8]
    Output partitioning: mmdsbx:HivePartitioningHandle{buckets=32, hiveTypes=[string, int]} [root_guid_0, integer(3)]
    ScanProject[table = mmdsbx:stg_connections_wo_duplicates]
        Layout: [root_guid_0:varchar, lvl1_root_guid:varchar, $hashvalue_8:bigint]
        $hashvalue_8 := combine_hash(bigint '0', COALESCE("$operator$hash_code"("root_guid_0"), 0))
        lvl1_root_guid := lvl1_root_guid:string:REGULAR
        root_guid_0 := root_guid:string:REGULAR
        Input: 7549086 rows (532.75MB), Filtered: 0.00%
```
Both tables are bucketed by the string column guid. The one on the probe side also has bucketing by another column which has a predicate on it and somehow we were able to create an output partitioning of bucketed column + constant on the build side scan.